examples/common/pigweed/rpc_services/Attributes.h

```diff
@@ -44,46 +45,177 @@ namespace rpc {
-            data = &request.data.data.data_int16;
+            write_status = writer.Put(kDataTag, data.data.data_int16);
             break;
         case chip_rpc_AttributeData_data_int32_tag:
-            data = &request.data.data.data_int32;
+            write_status = writer.Put(kDataTag, data.data.data_int32);
+            break;
+        case chip_rpc_AttributeData_data_single_tag:
+            write_status = writer.Put(kDataTag, data.data.data_single);
             break;
         case chip_rpc_AttributeData_data_bytes_tag:
-            data = &request.data.data.data_bytes;
+            write_status = writer.PutBytes(kDataTag, data.data.data_bytes.bytes, data.data.data_bytes.size);
             break;
-        case chip_rpc_AttributeData_data_single_tag:
-            data = &request.data.data.data_single;
+        case chip_rpc_AttributeData_data_string_tag:
+            write_status = writer.PutString(kDataTag, data.data.data_string);
             break;
         default:
             return pw::Status::InvalidArgument();
         }
-        RETURN_STATUS_IF_NOT_OK(
-            emberAfWriteAttribute(request.metadata.endpoint, request.metadata.cluster, request.metadata.attribute_id,
-                                  const_cast<uint8_t *>(static_cast<const uint8_t *>(data)), request.metadata.type));
+
+        if (write_status != CHIP_NO_ERROR)
+        {
+            ChipLogError(Support, "Failed to encode TLV data: %" CHIP_ERROR_FORMAT, write_status.Format());
+            return pw::Status::Internal();
+        }
+
+        VerifyOrReturnValue(writer.EndContainer(outer) == CHIP_NO_ERROR, pw::Status::Internal());
+        VerifyOrReturnValue(writer.Finalize() == CHIP_NO_ERROR, pw::Status::Internal());
+        result.Init(tempBuffer.data(), writer.GetLengthWritten());
+
+        VerifyOrReturnError(result.Next() == CHIP_NO_ERROR, pw::Status::Internal());
+        VerifyOrReturnError(result.EnterContainer(outer) == CHIP_NO_ERROR, pw::Status::Internal());
+
+        // This positions on the data element
+        VerifyOrReturnError(result.Next() == CHIP_NO_ERROR, pw::Status::Internal());
+
+        return result;
+    }
+
+    ::pw::Status Write(const chip_rpc_AttributeWrite & request, pw_protobuf_Empty & response)
+    {
+        app::ConcreteAttributePath path(request.metadata.endpoint, request.metadata.cluster, request.metadata.attribute_id);
+
+        DeviceLayer::StackLock lock;
+
+        // TODO: this assumes a singleton data model provider
+        app::DataModel::Provider * provider = app::InteractionModelEngine::GetInstance()->GetDataModelProvider();
+
+        app::DataModel::ServerClusterFinder serverClusterFinder(provider);
+        auto info = serverClusterFinder.Find(path);
+        if (!info.has_value())
+        {
+            return ::pw::Status::NotFound();
+        }
+
+        Access::SubjectDescriptor subjectDescriptor{ .authMode = chip::Access::AuthMode::kPase };
```

Comment:
Note that this makes it look like it's an EXTERNAL write. The goal of PW RPC is to make them look like the device did it itself. I might be useful to allow a new optional RPC argument that makes it "claim" PASE or "claim CASE on the last added fabric's fabric_idx" or such. FO rnow is it possible to use the internal write method (the one that would be used to internally set an attribute value such as when a device updates state with Ember" ?

Reply:
I am setting `write_request.operationFlags.Set(app::DataModel::OperationFlags::kInternal);` to flag this as an internal write.

Reply:
Setting a separate flag in SubjectDescriptor would have the benefit of it more clearly trickling down towards AAI or other implementations. If this is what you meant, I would rather do this as a followup as I am unsure about ripple effects of touching SubjectDescriptor (probably not much, however still it sounds like it should be a standalone change)

Reply:
Will do the update as a followup, to asses impact.